src/main/kotlin/org/rust/lang/core/psi/RsPsiTreeChangeListener.kt

```diff
@@ -0,0 +1,225 @@
+
+    sealed class ChildAdd(
+        override val file: PsiFile?,
+        val parent: PsiElement,
+        val child: PsiElement
+    ) : RsPsiTreeChangeEvent() {
+        class Before(file: PsiFile?, parent: PsiElement, child: PsiElement) : ChildAdd(file, parent, child)
+        class After(file: PsiFile?, parent: PsiElement, child: PsiElement) : ChildAdd(file, parent, child)
+
+        override fun toString(): String =
+            "ChildAdd.${javaClass.simpleName}(file=$file, parent=`${parent.text}`, child=`${child.text}`)"
+    }
+
+    sealed class ChildRemove(
+        override val file: PsiFile?,
+        val parent: PsiElement,
+        /** Invalid in [ChildRemove.After] */
+        val child: PsiElement
+    ) : RsPsiTreeChangeEvent() {
+        class Before(file: PsiFile?, parent: PsiElement, child: PsiElement) : ChildRemove(file, parent, child)
+        class After(file: PsiFile?, parent: PsiElement, child: PsiElement) : ChildRemove(file, parent, child)
+
+        override fun toString(): String =
+            "ChildRemove.${javaClass.simpleName}(file=$file, parent=`${parent.text}`, child=`${child.safeText}`)"
+    }
+
+    sealed class ChildReplace(
+        override val file: PsiFile?,
+        val parent: PsiElement,
+        /** Invalid in [ChildReplace.After] */
+        val oldChild: PsiElement,
+        val newChild: PsiElement
+    ) : RsPsiTreeChangeEvent() {
+        class Before(file: PsiFile?, parent: PsiElement, oldChild: PsiElement, newChild: PsiElement) : ChildReplace(file, parent, oldChild, newChild)
+        class After(file: PsiFile?, parent: PsiElement, oldChild: PsiElement, newChild: PsiElement) : ChildReplace(file, parent, oldChild, newChild)
+
+        override fun toString(): String =
+            "ChildReplace.${javaClass.simpleName}(file=$file, parent=`${parent.text}`, " +
+                "oldChild=`${oldChild.safeText}`, newChild=`${newChild.text}`)"
+    }
+
+    sealed class ChildMove(
+        override val file: PsiFile?,
+        val oldParent: PsiElement,
+        val newParent: PsiElement,
+        val child: PsiElement
+    ) : RsPsiTreeChangeEvent() {
+        class Before(file: PsiFile?, oldParent: PsiElement, newParent: PsiElement, child: PsiElement) : ChildMove(file, oldParent, newParent, child)
+        class After(file: PsiFile?, oldParent: PsiElement, newParent: PsiElement, child: PsiElement) : ChildMove(file, oldParent, newParent, child)
+
+        override fun toString(): String =
+            "ChildMove.${javaClass.simpleName}(file=$file, oldParent=`${oldParent.text}`, " +
+                "newParent=`${newParent.text}`, child=`${child.text}`)"
+
+
+    }
+
+    sealed class ChildrenChange(
+        override val file: PsiFile?,
+        val parent: PsiElement,
+        /**
+         * "generic change" event means that "something changed inside an element" and
+         * sends after all events for concrete PSI changes in the element.
+         */
+        val isGenericChange: Boolean
```

Comment:
Do we really need this property?
Currently, we always skip an event if `isGenericChange` is true. Maybe it's better not to create `ChildrenChange` events in such cases at all? It should reduce the count of allocated objects

Reply:
For debug purposes, I don't want to skip any events